File: sdk/prismer-cloud/typescript/src/encryption.ts
/**
 * Prismer SDK — E2E Encryption
 *
 * Industry-standard end-to-end encryption for IM messages.
 *
 * - Per-conversation symmetric key: AES-256-GCM
 * - Key exchange: ECDH P-256
 * - Master key derivation: PBKDF2-SHA256
 * - Key storage: Encrypted with master key
 * - Runtime: Web Crypto API (browser) / node:crypto (Node.js)
 *
 * The server only sees ciphertext — it cannot decrypt message content.
 */

// ============================================================================
// Crypto abstraction — works in both browser and Node.js
// ============================================================================

interface CryptoBackend {
  generateKeyPair(): Promise<{ publicKey: JsonWebKey; privateKey: JsonWebKey }>;
  deriveSharedSecret(privateKey: JsonWebKey, publicKey: JsonWebKey): Promise<ArrayBuffer>;
  deriveMasterKey(passphrase: string, salt: Uint8Array): Promise<ArrayBuffer>;
  encrypt(key: ArrayBuffer, plaintext: string): Promise<string>;
  decrypt(key: ArrayBuffer, ciphertext: string): Promise<string>;
  randomBytes(length: number): Uint8Array;
}

function getSubtleCrypto(): SubtleCrypto {
  if (typeof globalThis.crypto?.subtle !== 'undefined') {
    return globalThis.crypto.subtle;
  }
  // Node.js
  try {
    const { webcrypto } = require('node:crypto');
    return webcrypto.subtle;
  } catch {
    throw new Error('No SubtleCrypto available. Requires browser or Node.js 16+.');
  }
}

function getRandomValues(arr: Uint8Array): Uint8Array {
  if (typeof globalThis.crypto?.getRandomValues !== 'undefined') {
    return globalThis.crypto.getRandomValues(arr);
  }
  try {
    const { webcrypto } = require('node:crypto');
    return webcrypto.getRandomValues(arr);
  } catch {
    throw new Error('No crypto.getRandomValues available.');
  }
}

const subtle = (): SubtleCrypto => getSubtleCrypto();

const PBKDF2_ITERATIONS = 100_000;
const SALT_LENGTH = 16;
const IV_LENGTH = 12; // AES-GCM recommended
const KEY_LENGTH = 256; // AES-256

// ============================================================================
// E2E Encryption
// ============================================================================

/**
 * E2EEncryption manages per-conversation symmetric keys and
 * encrypts/decrypts message content using AES-256-GCM.
 *
 * Usage:
 *   const e2e = new E2EEncryption();
 *   await e2e.init('user-passphrase');
 *   const ciphertext = await e2e.encrypt('conv-123', 'Hello!');
 *   const plaintext = await e2e.decrypt('conv-123', ciphertext);
 */
export class E2EEncryption {
  private masterKey: CryptoKey | null = null;
  private keyPair: CryptoKeyPair | null = null;
  private sessionKeys = new Map<string, CryptoKey>(); // conversationId → AES key
  private salt: Uint8Array | null = null;

  /**
   * Initialize encryption with user passphrase.
   * Derives a master key via PBKDF2 and generates an ECDH key pair.
   *
   * @param passphrase - User passphrase for master key derivation
   * @param salt - Optional Base64-encoded salt. If omitted, a random 16-byte salt is generated.
   *               Store the salt (via exportSalt()) so you can re-derive the same master key later.
   */
  async init(passphrase: string, salt?: string): Promise<void> {
    // Use provided salt or generate random one
    this.salt = salt
      ? new Uint8Array(base64ToArrayBuffer(salt))
      : getRandomValues(new Uint8Array(SALT_LENGTH));

    // Derive master key from passphrase
    const passphraseKey = await subtle().importKey(
      'raw',
      new TextEncoder().encode(passphrase),
      'PBKDF2',
      false,
      ['deriveKey'],
    );

    this.masterKey = await subtle().deriveKey(
      {
        name: 'PBKDF2',
        salt: new Uint8Array(this.salt.buffer, this.salt.byteOffset, this.salt.byteLength).buffer as ArrayBuffer,
        iterations: PBKDF2_ITERATIONS,
        hash: 'SHA-256',
      },
      passphraseKey,
      { name: 'AES-GCM', length: KEY_LENGTH },
      false,
      ['encrypt', 'decrypt'],
    );

    // Generate ECDH key pair for key exchange
    this.keyPair = await subtle().generateKey(
      { name: 'ECDH', namedCurve: 'P-256' },
      true,
      ['deriveKey'],
    );
  }

  /**
   * Export the salt as Base64 string for persistent storage.
   * You must store this and pass it back to init() to re-derive the same master key.
   */
  exportSalt(): string {
    if (!this.salt) throw new Error('E2E not initialized. Call init() first.');
    return arrayBufferToBase64(this.salt.buffer as ArrayBuffer);
  }

  /**
   * Export public key for sharing with conversation peers.
   */
  async exportPublicKey(): Promise<JsonWebKey> {
    if (!this.keyPair) throw new Error('E2E not initialized. Call init() first.');
    return subtle().exportKey('jwk', this.keyPair.publicKey);
  }

  /**
   * Derive a shared session key for a conversation using ECDH.
   * Call this with each peer's public key.
   */
  async deriveSessionKey(conversationId: string, peerPublicKey: JsonWebKey): Promise<void> {
    if (!this.keyPair) throw new Error('E2E not initialized. Call init() first.');

    const importedPeerKey = await subtle().importKey(
      'jwk',
      peerPublicKey,
      { name: 'ECDH', namedCurve: 'P-256' },
      false,
      [],
    );

    const sessionKey = await subtle().deriveKey(
      { name: 'ECDH', public: importedPeerKey },
      this.keyPair.privateKey,
      { name: 'AES-GCM', length: KEY_LENGTH },
      false,
      ['encrypt', 'decrypt'],
    );

    this.sessionKeys.set(conversationId, sessionKey);
  }

  /**
   * Set a pre-shared session key for a conversation.
   * Useful when the key is exchanged out-of-band or derived from a group key.
   */
  async setSessionKey(conversationId: string, rawKey: ArrayBuffer): Promise<void> {
    const key = await subtle().importKey(
      'raw',
      rawKey,
      { name: 'AES-GCM', length: KEY_LENGTH },
      false,
      ['encrypt', 'decrypt'],
    );
    this.sessionKeys.set(conversationId, key);
  }

  /**
   * Generate a random session key for a conversation.
   * Returns the raw key bytes for sharing with peers.
   */
  async generateSessionKey(conversationId: string): Promise<ArrayBuffer> {
    const key = await subtle().generateKey(
      { name: 'AES-GCM', length: KEY_LENGTH },
      true,
      ['encrypt', 'decrypt'],
    );
    this.sessionKeys.set(conversationId, key);
    return subtle().exportKey('raw', key);
  }

  /**
   * Encrypt plaintext for a conversation.
   * Returns base64-encoded ciphertext with prepended IV.
   */
  async encrypt(conversationId: string, plaintext: string): Promise<string> {
    const key = this.sessionKeys.get(conversationId);
    if (!key) throw new Error(`No session key for conversation ${conversationId}. Call deriveSessionKey() first.`);

    const iv = getRandomValues(new Uint8Array(IV_LENGTH));
    const encoded = new TextEncoder().encode(plaintext);

    const ciphertext = await subtle().encrypt(
      { name: 'AES-GCM', iv: iv as BufferSource },
      key,
      encoded,
    );

    // Prepend IV to ciphertext and base64-encode
    const combined = new Uint8Array(iv.length + ciphertext.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(ciphertext), iv.length);

    return arrayBufferToBase64(combined.buffer);
  }

  /**
   * Decrypt ciphertext from a conversation.
   * Expects base64-encoded data with prepended IV.
   */
  async decrypt(conversationId: string, ciphertext: string): Promise<string> {
    const key = this.sessionKeys.get(conversationId);
    if (!key) throw new Error(`No session key for conversation ${conversationId}. Call deriveSessionKey() first.`);

    const combined = base64ToArrayBuffer(ciphertext);
    const iv = combined.slice(0, IV_LENGTH);
    const data = combined.slice(IV_LENGTH);

    const decrypted = await subtle().decrypt(
      { name: 'AES-GCM', iv: new Uint8Array(iv) },
      key,
      data,
    );

    return new TextDecoder().decode(decrypted);
  }

  /**
   * Check if a session key exists for a conversation.
   */
  hasSessionKey(conversationId: string): boolean {
    return this.sessionKeys.has(conversationId);
  }

  /**
   * Remove session key for a conversation.
   */
  removeSessionKey(conversationId: string): void {
    this.sessionKeys.delete(conversationId);
  }

  /**
   * Clear all keys and reset state.
   */
  destroy(): void {
    this.masterKey = null;
    this.keyPair = null;
    this.sessionKeys.clear();
    this.salt = null;
    this.messageCount = 0;
  }

  // ─── Pipeline Functions ──────────────────────────────────

  private messageCount = 0;
  private static readonly KEY_ROTATION_THRESHOLD = 1000;
  private static readonly KEY_ROTATION_INTERVAL_MS = 24 * 60 * 60 * 1000;
  private lastRotation = Date.now();

  /**
   * High-level encrypt-for-send pipeline.
   * Encrypts content, builds metadata, and handles key rotation.
   *
   * Returns { encryptedContent, metadata } ready to send.
   */
  async encryptForSend(
    conversationId: string,
    content: string,
  ): Promise<{ encryptedContent: string; metadata: Record<string, any> }> {
    if (!this.hasSessionKey(conversationId)) {
      throw new Error('No session key for this conversation. Call deriveSessionKey() first.');
    }

    // Check if key rotation is needed
    const needsRotation = this.shouldRotateKey();

    const encryptedContent = await this.encrypt(conversationId, content);

    this.messageCount++;

    return {
      encryptedContent,
      metadata: {
        encrypted: true,
        encryptionVersion: 1,
        ...(needsRotation && { keyRotationRequested: true }),
      },
    };
  }

  /**
   * High-level decrypt-on-receive pipeline.
   * Decrypts content and validates metadata.
   */
  async decryptOnReceive(
    conversationId: string,
    encryptedContent: string,
    metadata?: Record<string, any>,
  ): Promise<string> {
    if (!this.hasSessionKey(conversationId)) {
      throw new Error('No session key for this conversation. Call deriveSessionKey() first.');
    }

    return this.decrypt(conversationId, encryptedContent);
  }

  /**
   * High-level file encryption pipeline.
   */
  async encryptFile(
    conversationId: string,
    fileData: ArrayBuffer,
  ): Promise<{ encryptedData: string; metadata: Record<string, any> }> {
    const base64Data = arrayBufferToBase64(fileData);
    const encryptedData = await this.encrypt(conversationId, base64Data);
    return {
      encryptedData,
      metadata: {
        encrypted: true,
        encryptionVersion: 1,
        fileEncrypted: true,
      },
    };
  }

  /**
   * High-level file decryption pipeline.
   */
  async decryptFile(
    conversationId: string,
    encryptedData: string,
  ): Promise<ArrayBuffer> {
    const base64Data = await this.decrypt(conversationId, encryptedData);
    return base64ToArrayBuffer(base64Data);
  }

  /**
   * Check if key rotation is needed (1000 messages or 24 hours).
   */
  private shouldRotateKey(): boolean {
    if (this.messageCount >= E2EEncryption.KEY_ROTATION_THRESHOLD) {
      return true;
    }
    if (Date.now() - this.lastRotation >= E2EEncryption.KEY_ROTATION_INTERVAL_MS) {
      return true;
    }
    return false;
  }

  /**
   * Perform key rotation: generate new ECDH keypair and reset counters.
   * The caller is responsible for re-exchanging keys with peers.
   */
  async rotateKeys(): Promise<JsonWebKey> {
    // Generate new keypair
    this.keyPair = await subtle().generateKey(
      { name: 'ECDH', namedCurve: 'P-256' },
      false,
      ['deriveKey'],
    );

    // Reset counters
    this.messageCount = 0;
    this.lastRotation = Date.now();

    // Clear all session keys — peers need to re-derive
    this.sessionKeys.clear();

    return this.exportPublicKey();
  }
}

// ============================================================================
// Helpers
// ============================================================================

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  if (typeof btoa !== 'undefined') {
    // Browser
    const bytes = new Uint8Array(buffer);
    let binary = '';
    for (let i = 0; i < bytes.byteLength; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    return btoa(binary);
  }
  // Node.js
  return Buffer.from(buffer).toString('base64');
}

function base64ToArrayBuffer(base64: string): ArrayBuffer {
  if (typeof atob !== 'undefined') {
    // Browser
    const binary = atob(base64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes.buffer;
  }
  // Node.js
  const buf = Buffer.from(base64, 'base64');
  return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
}

File: sdk/prismer-cloud/rust/src/encryption.rs
//! E2E Encryption — AES-256-GCM with ECDH P-256 key exchange.
//!
//! Matches the TypeScript SDK's encryption implementation:
//! - Master key: PBKDF2-SHA256 (100k iterations) from passphrase
//! - Session keys: AES-256-GCM per conversation
//! - Key exchange: ECDH P-256

use aes_gcm::{
    aead::{Aead, KeyInit, OsRng},
    Aes256Gcm, Nonce,
};
use base64::{Engine as _, engine::general_purpose::STANDARD as BASE64};
use p256::{
    ecdh::EphemeralSecret,
    PublicKey,
};
use pbkdf2::pbkdf2_hmac;
use rand::RngCore;
use sha2::Sha256;
use std::collections::HashMap;

const SALT_LENGTH: usize = 16;

/// E2E encryption manager.
pub struct E2EEncryption {
    master_key: Option<[u8; 32]>,
    salt: Option<Vec<u8>>,
    session_keys: HashMap<String, [u8; 32]>,
    private_key: Option<EphemeralSecret>,
    public_key_bytes: Option<Vec<u8>>,
}

impl E2EEncryption {
    /// Create a new encryption manager.
    pub fn new() -> Self {
        Self {
            master_key: None,
            salt: None,
            session_keys: HashMap::new(),
            private_key: None,
            public_key_bytes: None,
        }
    }

    /// Initialize with a passphrase — derives master key via PBKDF2.
    ///
    /// If `salt_b64` is provided, it is decoded as Base64 and used as the PBKDF2 salt.
    /// Otherwise a random 16-byte salt is generated. Call `export_salt()` to persist it.
    pub fn init(&mut self, passphrase: &str) {
        self.init_with_salt(passphrase, None).expect("init with random salt should not fail");
    }

    /// Initialize with a passphrase and an explicit Base64-encoded salt.
    /// Call `export_salt()` to persist the salt for future re-derivation.
    pub fn init_with_salt(&mut self, passphrase: &str, salt_b64: Option<&str>) -> Result<(), String> {
        let salt_bytes = match salt_b64 {
            Some(s) => BASE64.decode(s).map_err(|e| format!("invalid base64 salt: {}", e))?,
            None => {
                let mut s = vec![0u8; SALT_LENGTH];
                OsRng.fill_bytes(&mut s);
                s
            }
        };
        let mut key = [0u8; 32];
        pbkdf2_hmac::<Sha256>(passphrase.as_bytes(), &salt_bytes, 100_000, &mut key);
        self.salt = Some(salt_bytes);
        self.master_key = Some(key);

        // Generate ECDH keypair
        let secret = EphemeralSecret::random(&mut OsRng);
        let public = PublicKey::from(&secret);
        self.public_key_bytes = Some(public.to_sec1_bytes().to_vec());
        self.private_key = Some(secret);
        Ok(())
    }

    /// Export the PBKDF2 salt as Base64 for persistent storage.
    pub fn export_salt(&self) -> Option<String> {
        self.salt.as_ref().map(|s| BASE64.encode(s))
    }

    /// Export the public key as base64 for sharing with peers.
    pub fn export_public_key(&self) -> Option<String> {
        self.public_key_bytes.as_ref().map(|bytes| BASE64.encode(bytes))
    }

    /// Set a pre-shared session key for a conversation.
    pub fn set_session_key(&mut self, conversation_id: &str, key: [u8; 32]) {
        self.session_keys.insert(conversation_id.to_string(), key);
    }

    /// Generate a random session key for a conversation.
    pub fn generate_session_key(&mut self, conversation_id: &str) -> [u8; 32] {
        let mut key = [0u8; 32];
        OsRng.fill_bytes(&mut key);
        self.session_keys.insert(conversation_id.to_string(), key);
        key
    }

    /// Encrypt plaintext for a conversation. Returns base64(IV + ciphertext).
    pub fn encrypt(&self, conversation_id: &str, plaintext: &str) -> Result<String, String> {
        let key = self.session_keys.get(conversation_id)
            .ok_or_else(|| format!("No session key for conversation: {}", conversation_id))?;

        let cipher = Aes256Gcm::new_from_slice(key)
            .map_err(|e| format!("Cipher init failed: {}", e))?;

        let mut iv = [0u8; 12];
        OsRng.fill_bytes(&mut iv);
        let nonce = Nonce::from_slice(&iv);

        let ciphertext = cipher.encrypt(nonce, plaintext.as_bytes())
            .map_err(|e| format!("Encryption failed: {}", e))?;

        // Prepend IV to ciphertext, then base64 encode
        let mut combined = Vec::with_capacity(12 + ciphertext.len());
        combined.extend_from_slice(&iv);
        combined.extend_from_slice(&ciphertext);

        Ok(BASE64.encode(&combined))
    }

    /// Decrypt base64(IV + ciphertext) for a conversation.
    pub fn decrypt(&self, conversation_id: &str, encrypted: &str) -> Result<String, String> {
        let key = self.session_keys.get(conversation_id)
            .ok_or_else(|| format!("No session key for conversation: {}", conversation_id))?;

        let combined = BASE64.decode(encrypted)
            .map_err(|e| format!("Base64 decode failed: {}", e))?;

        if combined.len() < 13 {
            return Err("Ciphertext too short".to_string());
        }

        let (iv, ciphertext) = combined.split_at(12);
        let cipher = Aes256Gcm::new_from_slice(key)
            .map_err(|e| format!("Cipher init failed: {}", e))?;
        let nonce = Nonce::from_slice(iv);

        let plaintext = cipher.decrypt(nonce, ciphertext)
            .map_err(|e| format!("Decryption failed: {}", e))?;

        String::from_utf8(plaintext).map_err(|e| format!("UTF-8 decode failed: {}", e))
    }

    /// Check if encryption is initialized.
    pub fn is_initialized(&self) -> bool {
        self.master_key.is_some()
    }

    /// Check if a conversation has a session key.
    pub fn has_session_key(&self, conversation_id: &str) -> bool {
        self.session_keys.contains_key(conversation_id)
    }
}

impl Default for E2EEncryption {
    fn default() -> Self {
        Self::new()
    }
}
